circuits/types/field/src/compare.rs

```diff
@@ -21,25 +21,52 @@ impl<E: Environment> Compare<Field<E>> for Field<E> {
-            if index != self_bits_be.len() - 1 {
-                are_previous_bits_equal &= self_bit.is_equal(other_bit);
-            }
+        if self.is_constant() && other.is_constant() {
+            Boolean::constant(self.eject_value() < other.eject_value())
+        } else if self.is_constant() {
+            // (For advanced users) this implementation saves us from instantiating 253 constants for
+            // the bits of `self`. The implementation in the `else` case invokes `to_bits_le` on
+            // `self` which would allocate 253 constants. Since `self` is constant, we can directly
+            // inspect its bits and construct an equivalent but reduced ternary expression.
+            self.eject_value().to_bits_le().into_iter().zip_eq(other.to_bits_le()).fold(
+                Boolean::constant(false),
+                |rest_is_less, (self_bit, other_bit)| {
+                    if self_bit {
+                        Boolean::ternary(&!&other_bit, &other_bit, &rest_is_less)
+                    } else {
+                        Boolean::ternary(&other_bit, &other_bit, &rest_is_less)
+                    }
+                },
+            )
+        } else if other.is_constant() {
+            // (For advanced users) this implementation saves us from instantiating 253 constants for
+            // the bits of `other`. The implementation in the `else` case invokes `to_bits_le` on
+            // `other` which would allocate 253 constants. Since `other` is constant, we can directly
+            // inspect its bits and construct an equivalent but reduced ternary expression.
+            self.to_bits_le().iter().zip_eq(other.eject_value().to_bits_le()).fold(
+                Boolean::constant(false),
+                |rest_is_less, (self_bit, other_bit)| {
+                    if other_bit {
+                        Boolean::ternary(&!self_bit, &!self_bit, &rest_is_less)
+                    } else {
+                        Boolean::ternary(self_bit, &!self_bit, &rest_is_less)
+                    }
+                },
+            )
+        } else {
+            // Zip `self.to_bits_le()` and `other.to_bits_le()` together and construct a check on the sequence of bit pairs.
+            // The bitwise check begins with the most-significant bit pair and ends with the least-significant bit pair.
+            // For each bit pair,
+            // - If `self_bit` and `other_bit` are different signs, then if `self_bit` is `true`, return false.
+            // - If `self_bit` and `other_bit` are different signs, then if `self_bit` is `false`, return true.
+            // - If `self_bit` and `other_bit` are the same sign, then check the following bits.
+            self.to_bits_le().iter().zip_eq(other.to_bits_le()).fold(
+                Boolean::constant(false),
+                |rest_is_less, (self_bit, other_bit)| {
+                    Boolean::ternary(&self_bit.bitxor(&other_bit), &other_bit, &rest_is_less)
```

Comment:
Can we combine the ternary with the bitxor here somehow? By, say, writing out the truth table.

Reply:
I'm not sure if there is a reduction for the ternary and bitxor.
I checked with a boolean function minimizer and the minimal form uses 3 AND gates and 2 OR gates.
I did minimize the logic for the cases where exactly one of the operands is constant.